File: books/mazes-in-go/generator/binary_tree.go
package generator

import (
	"math/rand"
	"mazes/maze"
)

func BinaryTree(grid *maze.Grid) {
	grid.EachCells(func(cell *maze.Cell) {
		if neighbor := chooseNeighbor(grid, cell); neighbor != nil {
			cell.Link(neighbor)
		}
	})
}

func chooseNeighbor(grid *maze.Grid, cell *maze.Cell) *maze.Cell {
	north := grid.North(cell)
	east := grid.East(cell)

	if north != nil && east != nil {
		if rand.Intn(2) == 0 {
			return north
		}
		return east
	}
	if north != nil {
		return north
	}
	if east != nil {
		return east
	}
	return nil
}

File: mazes-in-python/binary_tree.py
from __future__ import annotations
from random import Random
from operator import attrgetter

from grid import Grid


class BinaryTree:
    dirs = ['north', 'east', 'south', 'west']
    fs = [attrgetter(d) for d in dirs]

    def __init__(self, seed: int | None = None) -> None:
        self.rnd = Random(seed)

    def generate(self, direction: str, grid: Grid):
        index = self.dirs.index(direction)
        ln = len(self.fs)
        for cell in grid:
            cells = [self.fs[(index + i) % ln](cell) for i in [0, 1]]
            if ns := [v for v in cells if v]:
                if n := self.rnd.choice(ns):
                    cell.link(n)


def test_north():
    g = Grid(5, 5)
    tr = BinaryTree(6)
    tr.generate('north', g)
    # @formatter:off
    assert str(g) == \
"""
+---+---+---+---+---+
|                   |
+   +   +---+---+   +
|   |   |           |
+   +---+---+   +   +
|   |           |   |
+   +   +---+   +   +
|   |   |       |   |
+---+   +---+---+   +
|       |           |
+---+---+---+---+---+
"""
# @formatter:on


def test_east():
    g = Grid(5, 5)
    tr = BinaryTree(6)
    tr.generate('south', g)
    # @formatter:off
    assert str(g) == \
"""
+---+---+---+---+---+
|           |   |   |
+   +---+---+   +   +
|               |   |
+   +---+---+---+   +
|       |       |   |
+   +---+   +---+   +
|       |           |
+   +---+   +---+---+
|                   |
+---+---+---+---+---+
"""
# @formatter:on
